pkg/operator/operator_starter.go

```diff
@@ -0,0 +1,306 @@
+}
+
+type StandaloneStarter struct {
+	commonStarter
+}
+
+var _ OperatorStarter = &StandaloneStarter{}
+
+func NewStandaloneStarter(controllerConfig *controllercmd.ControllerContext) OperatorStarter {
+	ssr := &StandaloneStarter{}
+	ssr.controllerConfig = controllerConfig
+	return ssr
+}
+
+func (ssr *StandaloneStarter) StartOperator(ctx context.Context) error {
+	err := ssr.initClient(ctx)
+	if err != nil {
+		return err
+	}
+	err = ssr.commonStarter.CreateControllers()
+	if err != nil {
+		return err
+	}
+
+	err = ssr.commonStarter.getFeatureGate(ctx, ssr.eventRecorder)
+	if err != nil {
+		return err
+	}
+
+	countStorageClasses(ssr.commonClients)
+
+	csiDriverConfigs := ssr.populateConfigs(ssr.commonClients)
+	csiDriverController, _ := csidriveroperator.NewStandaloneDriverStarter(
+		ssr.commonClients,
+		ssr.featureGates,
+		resync,
+		ssr.versionGetter,
+		status.VersionForOperandFromEnv(),
+		ssr.eventRecorder,
+		csiDriverConfigs)
+	ssr.controllers = append(ssr.controllers, csiDriverController)
+
+	vsphereProblemDetector := vsphereproblemdetector.NewVSphereProblemDetectorStarter(
+		ssr.commonClients,
+		resync,
+		ssr.versionGetter,
+		status.VersionForOperandFromEnv(),
+		ssr.eventRecorder)
+	ssr.controllers = append(ssr.controllers, vsphereProblemDetector)
+
+	klog.Info("Starting the Informers.")
+
+	csoclients.StartInformers(ssr.commonClients, ctx.Done())
+
+	ssr.startControllers(ctx)
+	return fmt.Errorf("stopped")
+}
+
+func (ssr *StandaloneStarter) populateConfigs(clients *csoclients.Clients) []csioperatorclient.CSIOperatorConfig {
+	return []csioperatorclient.CSIOperatorConfig{
+		csioperatorclient.GetAWSEBSCSIOperatorConfig(false),
+		csioperatorclient.GetGCPPDCSIOperatorConfig(),
+		csioperatorclient.GetOpenStackCinderCSIOperatorConfig(clients, ssr.eventRecorder),
+		csioperatorclient.GetOVirtCSIOperatorConfig(clients, ssr.eventRecorder),
+		csioperatorclient.GetManilaOperatorConfig(clients, ssr.eventRecorder),
+		csioperatorclient.GetVMwareVSphereCSIOperatorConfig(),
+		csioperatorclient.GetAzureDiskCSIOperatorConfig(),
+		csioperatorclient.GetAzureFileCSIOperatorConfig(),
+		csioperatorclient.GetSharedResourceCSIOperatorConfig(false),
+		csioperatorclient.GetAlibabaDiskCSIOperatorConfig(),
+		csioperatorclient.GetIBMVPCBlockCSIOperatorConfig(),
+		csioperatorclient.GetPowerVSBlockCSIOperatorConfig(false),
+	}
+}
+
+type HyperShiftStarter struct {
+	commonStarter
+	guestKubeConfig string
+	mgmtClient      *csoclients.Clients
+}
+
+func NewHyperShiftStarter(controllerConfig *controllercmd.ControllerContext, guestKubeConfig string) OperatorStarter {
+	hsr := &HyperShiftStarter{}
+	hsr.controllerConfig = controllerConfig
+	hsr.guestKubeConfig = guestKubeConfig
+	return hsr
+}
+
+func (hsr *HyperShiftStarter) initClient(ctx context.Context) error {
+	controlPlaneNamespace := hsr.controllerConfig.OperatorNamespace
+
+	mgmtClients, err := csoclients.NewHypershiftMgmtClients(hsr.controllerConfig, controlPlaneNamespace, resync)
```

Comment:
Note to self: why NewHypershiftMgmtClients creates client for infrastructure in the mgmt cluster? The operator should read Infrastructure in the guest.

https://github.com/openshift/cluster-storage-operator/blob/9bdc393e16af5282154bb016ac38359d7c3b70c0/pkg/csoclients/csoclients.go#L162-L167

Reply:
I think we can do some more cleanup in a follow up. I was trying to keep this PR as close as possible to original code at least from logic perspective.